### PR TITLE
Reentrant parser cleanup

### PR DESCRIPTION
**Pull request by dnicolodi**
Some code simplification and cleanups on top of #491. These are the last patches I had from the large parser rework. The patch count is larger in the broken up PRs because I decided to split the changes differently, hopefully in an easier to review way. I actually dropped some useless changes (renames and the like...). I will push the parches with the move of the token semantic values construction to compiled code next.

### CODE REVIEW

**Comment by dnicolodi** [comment]
I didn't test the Bazel build and it fails miserably. Working on a fix.

File: beancount/parser/macros.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#define _CC_FUNC_03(FN, X, ...) FN(X); _CC_FUNC_02(FN, __VA_ARGS__)
+#define _CC_FUNC_04(FN, X, ...) FN(X); _CC_FUNC_03(FN, __VA_ARGS__)
+#define _CC_FUNC_05(FN, X, ...) FN(X); _CC_FUNC_04(FN, __VA_ARGS__)
+#define _CC_FUNC_06(FN, X, ...) FN(X); _CC_FUNC_05(FN, __VA_ARGS__)
+#define _CC_FUNC_07(FN, X, ...) FN(X); _CC_FUNC_06(FN, __VA_ARGS__)
+#define _CC_FUNC_08(FN, X, ...) FN(X); _CC_FUNC_07(FN, __VA_ARGS__)
+#define _CC_FUNC_09(FN, X, ...) FN(X); _CC_FUNC_08(FN, __VA_ARGS__)
+#define _CC_FUNC_10(FN, X, ...) FN(X); _CC_FUNC_09(FN, __VA_ARGS__)
+
+#define _CC_FUNC_SEQ(_01, _02, _03, _04, _05, _06, _07, _08, _09, _10, NAME, ...) NAME
```
Comment (blais) [inline beancount/parser/macros.h:15]: How does this work? I keep staring at it, don't grok it.
You call it below, passing __VA_ARGS__ as the first argument, I'm assuming this gets spliced into the call to __CC_FUNC_SEQ(), and then ... well that just eats up the arguments and expands to NAME... which is...what? 
This is simultaneously magic and beautiful yet incomprehensible.
Can you add a comment explaining how this is expanded?

File: beancount/parser/hashsrc.py [hunk 2]
```diff
@@ -18,6 +18,7 @@
 PARSER_SOURCE_FILES = [
     'lexer.l',
     'grammar.y',
+    'macros.h',
```
Comment (blais) [inline beancount/parser/hashsrc.py:21]: Do we need to add anything to setup.py?

File: beancount/parser/lexer_test.py [hunk 3]
```diff
@@ -429,13 +427,11 @@ def test_null_true_false(self, tokens, errors):
...
         TRUE FALSE NULL
         '''
-        # Note that this test contains an _actual_ newline, not an escape one as
-        # in the previous test. This should allow us to parse multiline strings.
         self.assertEqual([
             ('EOL', 2, '\n', None),
             ('BOOL', 2, 'TRUE', None),
             ('BOOL', 2, 'FALSE', None),
-            ('NULL', 2, 'NULL', None),
+            ('NONE', 2, 'NULL', None),
```
Comment (blais) [inline beancount/parser/lexer_test.py:434]: Why this change? I'm calling them "NULL" in Beancount.

File: beancount/parser/hashsrc.py (hunk 2 above)
Comment (dnicolodi) [reply beancount/parser/hashsrc.py:21]: If you mean to include the new file in the distribution, I don't think so: we use a wildcard expansion in MANIFEST.in

File: beancount/parser/lexer_test.py (hunk 3 above)
Comment (dnicolodi) [reply beancount/parser/lexer_test.py:434]: The token is called NONE in the grammar definition https://github.com/beancount/beancount/blob/master/beancount/parser/grammar.y#L211 , its name is translated into "NULL" in the function that this commits drops https://github.com/beancount/beancount/blob/master/beancount/parser/grammar.y#L941 I didn't fee like this is important enough to rename the token or special case it in the `token_to_string()` function. Either can be done if you think the name NULL is more appropriate.

File: beancount/parser/macros.h (hunk 1 above)
Comment (dnicolodi) [reply beancount/parser/macros.h:15]: Indeed it takes a while to see how this works, but once you see it, it is obvious. It would be easier to find an article that describes the technique, but I haven't found one in a rapid search. I'll either add some comments in the form of an explanation or a link to one.

Comment (blais) [reply beancount/parser/macros.h:15]: Gimme some comment on this and we're good to go. Ideally a link...

File: beancount/parser/lexer_test.py (hunk 3 above)
Comment (blais) [reply beancount/parser/lexer_test.py:434]: Fair enough.

**Comment by blais** [comment]
The error from CI will show you that the quoting from your macro tricks is probably not working:

```
beancount/parser/hashsrc.py:62
  /home/travis/build/beancount/beancount/beancount/parser/hashsrc.py:62: UserWarning: The Beancount parser C extension module is out-of-date ('PARSER_SOURCE_HASH' != '86a5881b897b0b11eb4e001ca1ebd274'). You need to rebuild.
    warnings.warn(

```

File: beancount/parser/macros.h (hunk 1 above)
Comment (dnicolodi) [reply beancount/parser/macros.h:15]: I didn't find a link to an explanation I liked. Thus I wrote one. Let me know if it is understandable.

**Comment by dnicolodi** [comment]
> The error from CI will show you that the quoting from your macro tricks is probably not working:

That's what you get for trying to be too smart... I think a more elegant way could have been found, but I decided to keep the code simple. Please have a look.
